File: milliseconds/golang/services/millisecond_service.go
package services

import (
	"context"
	modelrequests "note-golang-millisecond/models/requests"
	modelresponses "note-golang-millisecond/models/responses"
	"time"
)

type MillisecondService interface {
	GetByGMTPlus8(ctx context.Context, millisecondRequest modelrequests.MillisecondRequest) (millisecondResponse modelresponses.MillisecondResponse)
	GetByGMTMinus8(ctx context.Context, millisecondRequest modelrequests.MillisecondRequest) (millisecondResponse modelresponses.MillisecondResponse)
}

type millisecondService struct {
}

func NewMillisecondService() MillisecondService {
	return &millisecondService{}
}

func (service *millisecondService) GetByGMTPlus8(ctx context.Context, millisecondRequest modelrequests.MillisecondRequest) (millisecondResponse modelresponses.MillisecondResponse) {
	// Offset -8 jam = -8 * 3600 detik = -28800
	gmtplus8 := time.FixedZone("GMT+08:00", 8*60*60)

	t := time.Date(millisecondRequest.Year, time.Month(millisecondRequest.Month), millisecondRequest.Date, millisecondRequest.Hour, millisecondRequest.Minute, millisecondRequest.Second, 0, gmtplus8)
	millisecondResponse.Datetime = t.String()
	millisecondResponse.Millisecond = t.UnixMilli()
	millisecondResponse.Add1Hour = t.Add(1 * time.Hour).String()
	millisecondResponse.Add1HourMillisecond = t.Add(1 * time.Hour).UnixMilli()
	return
}

func (service *millisecondService) GetByGMTMinus8(ctx context.Context, millisecondRequest modelrequests.MillisecondRequest) (millisecondResponse modelresponses.MillisecondResponse) {
	// Offset -8 jam = -8 * 3600 detik = -28800
	gmtMinus8 := time.FixedZone("GMT-08:00", -8*60*60)
	t := time.Date(millisecondRequest.Year, time.Month(millisecondRequest.Month), millisecondRequest.Date, millisecondRequest.Hour, millisecondRequest.Minute, millisecondRequest.Second, 0, gmtMinus8)
	millisecondResponse.Datetime = t.String()
	millisecondResponse.Millisecond = t.UnixMilli()
	millisecondResponse.Add1Hour = t.Add(1 * time.Hour).String()
	millisecondResponse.Add1HourMillisecond = t.Add(1 * time.Hour).UnixMilli()
	return
}

File: milliseconds/rust/src/services/millisecond_service.rs
use chrono::{DateTime, Duration, FixedOffset, NaiveDate, TimeZone};

use crate::models::{requests::millisecond_request::MillisecondRequest, responses::millisecond_response::MillisecondResponse};

pub trait MillisecondService {
    async fn get_by_gmt_plus8(&self, millisecond_request: MillisecondRequest) -> MillisecondResponse;
    async fn get_by_gmt_minus8(&self, millisecond_request: MillisecondRequest) -> MillisecondResponse;
}

pub struct MillisecondServiceImpl {

}

impl MillisecondServiceImpl {
    pub fn new() -> MillisecondServiceImpl {
        MillisecondServiceImpl {  }
    }
}

impl MillisecondService for MillisecondServiceImpl {
    async fn get_by_gmt_plus8(&self, millisecond_request: MillisecondRequest) -> MillisecondResponse {
        let offset_plus8 = FixedOffset::east_opt(8 * 60 * 60).unwrap();
        let naive_dt = NaiveDate::from_ymd_opt(millisecond_request.year, millisecond_request.month, millisecond_request.date).unwrap().and_hms_opt(millisecond_request.hour, millisecond_request.minute, millisecond_request.second).unwrap();
        let dt_plus8: DateTime<FixedOffset> = offset_plus8.from_local_datetime(&naive_dt).unwrap();
        let dt_plus8_add1_hour = dt_plus8 + Duration::hours(1);
        MillisecondResponse::set_millisecond_response(dt_plus8.to_string(), dt_plus8.timestamp_millis(), dt_plus8_add1_hour.to_string(), dt_plus8_add1_hour.timestamp_millis())
    }

    async fn get_by_gmt_minus8(&self, millisecond_request: MillisecondRequest) -> MillisecondResponse {
        let offset_minus8 = FixedOffset::west_opt(8 * 60 * 60).unwrap();
        let naive_dt = NaiveDate::from_ymd_opt(millisecond_request.year, millisecond_request.month, millisecond_request.date).unwrap().and_hms_opt(millisecond_request.hour, millisecond_request.minute, millisecond_request.second).unwrap();
        let dt_plus8: DateTime<FixedOffset> = offset_minus8.from_local_datetime(&naive_dt).unwrap();
        let dt_minus8_add1_hour = dt_plus8 + Duration::hours(1);
        MillisecondResponse::set_millisecond_response(dt_plus8.to_string(), dt_plus8.timestamp_millis(), dt_minus8_add1_hour.to_string(), dt_minus8_add1_hour.timestamp_millis())
    }
}